### PR TITLE
Initial integration of ZenDNN as backend into PyTorch

### PR DESCRIPTION
**Pull request by ratan-prasad**
This PR covers following
- Added ZenDNN as third-party library
    - ZenDNN is a submodule for pytorch
    - modified pytorch build infra to build with ZenDNN when USE_ZENDNN is set
      to 1
    - pytorch build reverts to default behavior if USE_ZENDNN environmental
      variable is set to zero or unset
- Added AOCL-blis as  dependency of ZenDNN
    - AOCL-blis is a submodule for pytorch
    - AOCL-blis is built if USE_ZENDNN is set to 1
- Added ZenDNN as backend to pytorch
    - added support for user level control to disable or enable zendnn backend
- Added following changes
    - added zendnn layout and ZendnnCPU dispatch key.
    - tensor conversion between dense and zendnn layouts
    - zendnn convolution as a new ConvBackend
        - when built with zendnn (USE_ZENDNN=1) and gradient calculations are
          disabled (torch.no_grad) convolution is computed using ZenDNN library)
	- backward propagation is not supported.
	- support is only added for fp32 data type
- Added few unit tests for pytorch + zendnn

Signed-off-by: Ratan Prasad <ratan.prasad2@amd.com>

Fixes #ISSUE_NUMBER

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### CODE REVIEW

**Comment by malfet** [comment]
@ratan-prasad can you please share some performance numbers in PR description? Also, what is the expected binary footprint increase as result of this change?

**Comment by ratan-prasad** [comment]
> @ratan-prasad can you please share some performance numbers in PR description? Also, what is the expected binary footprint increase as result of this change?
@malfet This PR contains only one op integrated to validate the procedure ahead of submitting the remaining PRs integrating the rest of the ZenDNN ops. To showcase the potential uplift using ZenDNN vs. Intel MKL-DNN on AMD CPU, we have benchmarked our full integrated version on PT v1.11. 
PyTorch v1.11 Throughput(images/sec) BS:640, Milan 64C NPS4

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<link rel=themeData
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">





</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>




Models | Improvement
-- | --
ResNet50 | 6.49%
GoogleNet | 118.83%
vgg11 | 18.17%
Inception-v4 | 69.96%
Alexnet | 17.99%




</body>

</html>

PyTorch v1.11 Latency(ms) BS:1 Milan 64C NPS4

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<link rel=themeData
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">





</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>




Models | Improvement
-- | --
ResNet50 | 6.49%
GoogleNet | 118.83%
vgg11 | 18.17%
Inception-v4 | 69.96%
Alexnet | 17.99%




</body>

</html>


Wheel file size comparison with and without ZenDNN build as below
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<link rel=themeData
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">





</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>




Build | PR1 (size in MB)
-- | --
Default build | 105
Build with ZenDNN | 107




</body>

</html>

**Comment by huiguoo** [comment]
@ratan-prasad Are the performance numbers exactly the same for batch size 1 and batch size 640? Have you tried other batch sizes?

**Comment by ratan-prasad** [comment]
> @ratan-prasad Are the performance numbers exactly the same for batch size 1 and batch size 640? Have you tried other batch sizes?

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<link rel=themeData
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">



</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>


<p class=MsoNormal><span style='background:yellow;mso-highlight:yellow'>Sorry
for late reply. We have benchmarked for other batch sizes also. There was some
typo for batch size 1 in previous reply. We observed that conda installation of
PyTorch v1.11.0 performs better than pip based installation for oneDNN. Hence
we have regenerated benchmarking data. Please find the updated improvements
info as below.</span><o:p></o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>Please find the benchmarking details below<o:p></o:p></p>

<p class=MsoNormal>PyTorch v1.11 Throughput(images/sec) <b>BS:640</b>, Milan
64C, NPS4<o:p></o:p></p>



Models | Improvement
-- | --
ResNet50 | 11.09%
GoogleNet | 113.91%
vgg11 | 17.57%
Inception-v4 | 70.13%
Alexnet | 5.05%




</body>

</html>

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<link rel=themeData
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">





</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>


<p class=MsoListParagraph style='margin-left:0in'>PyTorch v1.11
Throughput(images/sec) <b>BS:128</b>, Milan 64C, NPS4<o:p></o:p></p>



Models | Improvement
-- | --
ResNet50 | 4.14%
GoogleNet | 107.84%
vgg11 | 13.66%
Inception-v4 | 63.66%
Alexnet | 14.61%




</body>

</html>

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<link rel=themeData
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">





</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>


<p class=MsoListParagraph style='margin-left:0in'>PyTorch v1.11
Throughput(images/sec) <b>BS:1</b>, Milan 64C, NPS4<o:p></o:p></p>



Models | Improvement
-- | --
ResNet50 | 12.38%
GoogleNet | 66.27%
vgg11 | 154.63%
Inception-v4 | 47.40%
Alexnet | 324.45%




</body>

</html>

**Review by malfet — CHANGES_REQUESTED** [review]
Proposed solution makes `zennDNN` and `BLIS` a mandatory submodules of PyTorch even though dependence itself will be used very sporadically (as currently there are no CI/CD support)

Can we please solve mutual exclusivity issue or modify the PR to make those optional dependencies(checked out only if USE_ZENDNN is set to true) and reuse the same `DNNCPU` dispatch key

**Comment by ratan-prasad** [comment]
> Proposed solution makes `zennDNN` and `BLIS` a mandatory submodules of PyTorch even though dependence itself will be used very sporadically (as currently there are no CI/CD support)
> 
> Can we please solve mutual exclusivity issue or modify the PR to make those optional dependencies(checked out only if USE_ZENDNN is set to true) and reuse the same `DNNCPU` dispatch key

Should we modify “CMakeLists.txt” or “cmake/Dependencies.cmake” so that ZenDNN and BLIS repositories will be downloaded during setup.py execution ? (when USE_ZENDNN=1 is exported).

Note: We are also looking forward to enable CI runs for ZenDNN builds.

File: CMakeLists.txt [hunk 1]
```diff
@@ -128,9 +128,11 @@ endif()
 
 set(CPU_AARCH64 OFF)
 set(CPU_INTEL OFF)
+set(CPU_AMD OFF)
```
[the remark this answers is not in the document]
Comment (ratan-prasad) [reply CMakeLists.txt:131]: Thanks for the suggestion. We wanted to keep the existing code untouched as much as possible.
We will submit this correction in another patch on the same pull request.

File: CMakeLists.txt [hunk 2]
```diff
@@ -304,6 +313,16 @@ cmake_dependent_option(
     USE_MKLDNN_CBLAS "Use CBLAS in MKLDNN" OFF
     "USE_MKLDNN" OFF)
 option(USE_STATIC_MKL "Prefer to link with MKL statically (Unix only)" OFF)
+
+
+cmake_dependent_option(
+  USE_ZENDNN "Use ZENDNN. Only available on x86 and x86_64." "${CPU_AMD}"
```
[the remark this answers is not in the document]
Comment (ratan-prasad) [reply CMakeLists.txt:319]: Thanks for the suggestion. We will submit this correction in another patch on the same pull request.

File: c10/core/Backend.h [hunk 3]
```diff
@@ -51,6 +51,7 @@ enum class Backend {
   QuantizedXPU,
   Undefined,
   MkldnnCPU,
+  ZendnnCPU,
```
[the remark this answers is not in the document]
Comment (ratan-prasad) [reply c10/core/Backend.h:54]: If we use same dispatch key we are anticipating some challenges in ZenDNN integration.

E.g.
How to update “aten/src/ATen/native/native_functions.yaml” for mkldnn and zendnn integrations?

Currently a relu function entry in “aten/src/ATen/native/native_functions.yaml” looks as follows

func: relu(Tensor self) -> Tensor
  device_check: NoCheck   # TensorIterator
  variants: function, method
  dispatch:
    CPU, CUDA: relu
    MkldnnCPU: mkldnn_relu
    ZendnnCPU: zendnn_relu
    QuantizedCPU: relu_quantized_cpu

How to make the same dnnCPU dispatch key to point to mkldnn_relu or zendnn_relu based on CMAKE Level flags? Can you please point us to any example for this?

Note:- ATEN calls mkldnn using IDEEP API and IDEEP library. ATEN will be calling ZenDNN directly using ZenDNN API without IDEEP API/library

File: c10/core/TensorImpl.h [hunk 4]
```diff
@@ -966,10 +973,10 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
...
     // strided case first.
     // This keyset must also be kept in sync with the logic in
-    // is_sparse() / is_sparse_csr() / is_mkldnn()
-    constexpr auto sparse_and_sparsecsr_and_mkldnn_ks =
-        c10::sparse_ks | c10::sparse_csr_ks | c10::mkldnn_ks;
-    if (!key_set_.has_any(sparse_and_sparsecsr_and_mkldnn_ks)) {
+    // is_sparse() / is_sparse_csr() / is_mkldnn() / is_zendnn()
+    constexpr auto sparse_and_sparsecsr_and_mkldnn_ks_and_zendnn_ks =
+        c10::sparse_ks | c10::sparse_csr_ks | c10::mkldnn_ks | c10::zendnn_ks;
+    if (!key_set_.has_any(sparse_and_sparsecsr_and_mkldnn_ks_and_zendnn_ks)) {
```
[the remark this answers is not in the document]
Comment (ratan-prasad) [reply c10/core/TensorImpl.h:979]: Same answer as previous.

File: c10/core/Backend.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (xw285cornell) [reply c10/core/Backend.h:54]: @malfet I think one issue with sharing the same key is that we have to build separate packages for AMD and Intel which isn't ideal. Ideally we can use the same package and at runtime pick different backend. 

@ratan-prasad curious about the mutually exclusive part: why is that the case? Is it symbol conflict or something?

[the remark this answers is not in the document]
Comment (ratan-prasad) [reply c10/core/Backend.h:54]: @xw285cornell Yes, ZenDNN uses AMD-BLIS (a CBLAS API library) for optimal matrix operations on AMD hardware. AMD-BLIS has symbol clashes with MKL. That's why we are planning for mutually exclusive builds.
Can you help us with sample code or existing similar integrations where same dispatch key shared between two different libraries?

**Comment by b0noI** [comment]
is there RFC for this? If no we probably should proceed with the review but for the next big changes like this we probably should have RFC first. If there is RFC reviewed and approved already please link it.

**Comment by Willian-Zhang** [comment]
Very excited to see this is going on rapidly. Just saw Zen 4 EPYC release today and huge performance boost along with it, can't wait to get one myself playing around.

**Comment by ratan-prasad** [comment]
HI @xw285cornell, @malfet, @bdhirsh, @robieta can you please help us in removing the **Stale** tag?

**Comment by malfet** [comment]
@ratan-prasad i'm not sure, why it should be removed. PR in it's current form reached a stalemate and as such can not be merged.

**Comment by jgong5** [comment]
I took a quick look at the ZenDNN code: https://github.com/amd/ZenDNN. It seems a lot of code was just a clone of oneDNN (a.k.a. MKLDNN). Have you considered to contribute optimizations to oneDNN directly? This can make PyTorch side simpler. You don't have to clone the oneDNN integration again here. cc @vpirogov

**Comment by naveenthangudu** [comment]
> @ratan-prasad i'm not sure, why it should be removed. PR in it's current form reached a stalemate and as such can not be merged.

Can you please let us what can be done to bypass this stalemate?

**Comment by soumith** [comment]
what would definitely remove the stalemate is to upstream your patches to OneDNN.
I didn't realize ZenDNN is a fork of OneDNN with a bunch of duplicated code -- but after @jgong5 pointed it out, I went and looked and it seems true.
It seems completely wasteful and bug-prone to fork it.

If you wish to continue maintaining the fork because of company dynamics, etc., then as Nikita stated above, this PR needs to:
1. be rebased onto latest `master` commit
2. made a completely optional cmake flag (and no submodules should be committed to the pytorch repo), as @ratan-prasad described in this comment: https://github.com/pytorch/pytorch/pull/76242#issuecomment-1179977899
3. Because ZenDNN is a fork of OneDNN with minor API changes, similar to how ROCm does a regex of `s/cuda/hip/`, the ZenDNN patch can regex `s/mkldnn/zendnn/` to keep the changes to PyTorch core minimal (instead of having ~3000 lines enter PyTorch core that we'll have to maintain)

**Comment by naveenthangudu** [comment]
@malfet @soumith  Thank you for inputs to avoid stalemate on ZenDNN integration.

>     1. be rebased onto latest `master` commit
With substitution approach, integration changes will be different compared to this PR. Hence, we are planning to raise new PR. 

>     2. made a completely optional cmake flag (and no submodules should be committed to the pytorch repo), as @ratan-prasad described in this comment: [Initial integration of ZenDNN as backend into PyTorch #76242 (comment)](https://github.com/pytorch/pytorch/pull/76242#issuecomment-1179977899)
Can we use cmake API(“ExternalProject_Add”) for downloading ZenDNN repos?
 
>     3. Because ZenDNN is a fork of OneDNN with minor API changes, similar to how ROCm does a regex of `s/cuda/hip/`, the ZenDNN patch can regex `s/mkldnn/zendnn/` to keep the changes to PyTorch core minimal (instead of having ~3000 lines enter PyTorch core that we'll have to maintain)
currently working on a POC of this approach. Should I create an RFC for this approach?

**Comment by gottbrath** [comment]
> > 3. Because ZenDNN is a fork of OneDNN with minor API changes, similar to how ROCm does a regex of `s/cuda/hip/`, the ZenDNN patch can regex `s/mkldnn/zendnn/` to keep the changes to PyTorch core minimal (instead of having ~3000 lines enter PyTorch core that we'll have to maintain)
> > ```
> 
> currently working on a POC of this approach. Should I create an RFC for this approach?

I think an POC and RFC is a great idea for something of this magnitude/potential complexity.

**Comment by gottbrath** [comment]
see https://github.com/pytorch/rfcs/pull/52

**Pull request closed without merging**